src/main/java/org/marketdesignresearch/mechlib/mechanism/auctions/cca/CCAuction.java

```diff
@@ -54,57 +27,37 @@ public CCAuction(Domain domain, OutcomeRuleGenerator outcomeRuleGenerator) {
     }
 
     public CCAuction(Domain domain, OutcomeRuleGenerator mechanismType, Prices currentPrices) {
-        this(domain, mechanismType, false);
-        this.currentPrices = currentPrices;
+        super(domain,mechanismType,new CCAClockPhase(currentPrices));
+        setMaxRounds(100);
     }
 
     public CCAuction(Domain domain, OutcomeRuleGenerator mechanismType, boolean proposeStartingPrices) {
-        super(domain, mechanismType);
-        this.supplementaryRounds = new ArrayList<>();
+        super(domain, mechanismType, new CCAClockPhase(domain, proposeStartingPrices));
         setMaxRounds(100);
-        if (proposeStartingPrices) {
-            this.currentPrices = getDomain().proposeStartingPrices();
-        } else {
-            this.currentPrices = new LinearPrices(getDomain().getGoods());
-        }
     }
 
-    @PersistenceConstructor
-    private CCAuction(Domain domain, OutcomeRuleGenerator outcomeRuleGenerator, Prices currentPrices, PriceUpdater priceUpdater, List<SupplementaryRound> supplementaryRounds, LinkedList<SupplementaryRound> supplementaryRoundQueue, boolean clockPhaseCompleted) {
-        super(domain, outcomeRuleGenerator);
-        this.currentPrices = currentPrices;
-        this.priceUpdater = priceUpdater;
-        this.supplementaryRounds = supplementaryRounds;
-        this.supplementaryRoundQueue = supplementaryRoundQueue;
-        this.clockPhaseCompleted = clockPhaseCompleted;
+    public void addSupplementaryRound(SupplementaryPhase supplementaryRound) {
+        this.addAuctionPhase(supplementaryRound);
     }
-
-    @Override
-    public int allowedNumberOfBids() {
-        if (!clockPhaseCompleted) return 1;
-        SupplementaryRound next = supplementaryRoundQueue.peek();
-        return next == null ? 0 : next.getNumberOfSupplementaryBids();
-    }
-
-    public ImmutableList<SupplementaryRound> getSupplementaryRounds() {
-        return ImmutableList.copyOf(supplementaryRounds);
-    }
-
-    public void addSupplementaryRound(SupplementaryRound supplementaryRound) {
-        supplementaryRounds.add(supplementaryRound);
-        supplementaryRoundQueue.add(supplementaryRound);
+    
+    public void setPriceUpdater(PriceUpdater updater) {
+    	((CCAClockPhase)this.phases.get(0)).setPriceUpdater(updater);
```

Comment:
Does that mean that a price updater could be changed during the auction?

Reply:
This is actually taken from the original implementation of the CCA. To me it does not hurt to offer this option to a bidder.
The only downside in offering this method is that a rollback to a certain round would might change the auction.
It could be easily changed setting the PriceUpdater only via the constructor.

Reply:
I leave this decision to you - the downside you're mentioning hints at this design flaw (that was obviously there before, no blame here). I don't think it's of priority, but maybe it would be better to set it via constructor...

Reply:
I changed that. The PriceUpdater is now set via constructor